CV32E40Pv2 220 CORE-V SIMD Builtins Version 1

NandniJamnadas opened the pull request:
* `gcc/common/config/riscv/riscv-common.cc`: Added XCOREVSIMD extensions. Added riscv_xcorev_flags to SIMD extensions.
* `gcc/config/riscv/riscv-opts.h`: Added corresponding MASK and TARGET macros for XCOREVSIMD.
* `gcc/config/riscv/constraints.md`: Created
    * 6-bit signed immediate constraint "CV6"
    * 6-bit unsigned immediate constraint "CS6"
    * 4-bit shifting immediate constraint:
      * constant integer value "0" constraint "CF0"
      * constant integer value "1" constraint "CF1"
      * constant integer value "2" constraint "CF2"
      * constant integer value "3" constraint "CF3"
* `gcc/config/riscv/corev.def`: Defined 142 SIMD builtins for 220 SIMD instructions with RISCV_BUILTIN.
* `gcc/config/riscv/corev.md`:
  * Machine description pattern for all 142 SIMD builtins with UNSPEC version.
  * Included `define_c_enum "unspec"` for CORE-V specific unspec macros.
* `gcc/config/riscv/predicates.md`: Created
    * 6-bit signed immediate predicate for constraint "CV6"
    * 6-bit unsigned immediate predicate for constraint "CS6"
    * 4-bit unsigned immediate predicate for constraint "CF0,CF1,CF2,CF3"
* `gcc/config/riscv/riscv-builtins.cc`: Created SIMD Avaliability Predicate, RISCV_FTYPE_NAME2, RISCV_FTYPE_ATYPES2. Added
  corresponding argument types for builtins.
* `gcc/config/riscv/riscv-ftypes.def`: Added corresponding DEF_RISCV_FTYPE return type and argument types for builtins.
* `gcc/testsuite/gcc.target/riscv/cv-simd-*`: 158 Compilation Tests and 1 MARCH test: `cv-xcorev-simd-march-compile-1.c`.
* `gcc/doc/extend.texi`: Updated documentation to include
	  CORE-V SIMD Builtins.

NandniJamnadas committed:
CV32E40Pv2 CORE-V SIMD Extension Only

    * gcc/common/config/riscv/riscv-common.cc: Added XCOREVSIMD
      extensions. Added riscv_xcorev_flags to SIMD extensions.
    * gcc/config/riscv/riscv-opts.h: Added corresponding MASK and
      TARGET macros for XCOREVSIMD.

NandniJamnadas commented:
GCC testing, RISC-V Target ONLY:

Category | Previous | With commit | Delta
-- | -- | -- | --
Expected passes | 3681 | 7980 | +4299
Unexpected failures | 53 | 53 | -
Unexpected successes | - | - | -
Expected failures | 6 | 6 | -
Unresolved testcases | 6 | 6 | -
Unsupported tests | 233 | 233 | -

jeremybennett commented on gcc/config/riscv/corev.def:
```diff
@@ -20,3 +20,225 @@ RISCV_BUILTIN (cv_mac_macuRN,    "cv_mac_macuRN",     RISCV_BUILTIN_DIRECT, RISC
+RISCV_BUILTIN (cv_simd_add_sc_h_si,         "cv_simd_add_sc_h",         RISCV_BUILTIN_DIRECT, RISCV_USI_FTYPE_USI_USI, cvsimd),
+RISCV_BUILTIN (cv_simd_add_sc_b_si,         "cv_simd_add_sc_b",         RISCV_BUILTIN_DIRECT, RISCV_USI_FTYPE_USI_USI, cvsimd),
+RISCV_BUILTIN (cv_simd_add_sci_h_si,        "cv_simd_add_sci_h",        RISCV_BUILTIN_DIRECT, RISCV_USI_FTYPE_USI_SI, cvsimd),
+RISCV_BUILTIN (cv_simd_add_sci_b_si,        "cv_simd_add_sci_b",        RISCV_BUILTIN_DIRECT, RISCV_USI_FTYPE_USI_SI, cvsimd),
```
As discussed off line, all the `sci` variants are going to be replaced by patterns that automatically decide.

jeremybennett replied:
Resolved in the latest version

jeremybennett replied:
Resolved by newer commits.

jeremybennett commented on gcc/config/riscv/corev.def:
```diff
@@ -20,3 +20,225 @@ RISCV_BUILTIN (cv_mac_macuRN,    "cv_mac_macuRN",     RISCV_BUILTIN_DIRECT, RISC
 RISCV_BUILTIN (cv_mac_machhsRN,  "cv_mac_machhsRN",   RISCV_BUILTIN_DIRECT, RISCV_SI_FTYPE_SI_SI_SI_UQI,      cvmac),
+
+// xcoresimd
+RISCV_BUILTIN (cv_simd_add_h_si,            "cv_simd_add_h",            RISCV_BUILTIN_DIRECT, RISCV_USI_FTYPE_USI_USI, cvsimd),
```
Shouldn't this (and `cv_simd_sub_h`) be of type `RISCV_USI_FTYPE_USI_USI_UQI`

jeremybennett replied:
Resolved by newer commits.

jeremybennett commented on gcc/config/riscv/corev.def:
```diff
@@ -20,3 +20,225 @@ RISCV_BUILTIN (cv_mac_macuRN,    "cv_mac_macuRN",     RISCV_BUILTIN_DIRECT, RISC
+RISCV_BUILTIN (cv_simd_and_sci_b_si,        "cv_simd_and_sci_b",        RISCV_BUILTIN_DIRECT, RISCV_USI_FTYPE_USI_SI, cvsimd),
+RISCV_BUILTIN (cv_simd_abs_h_si,            "cv_simd_abs_h",            RISCV_BUILTIN_DIRECT, RISCV_USI_FTYPE_USI, cvsimd),
+RISCV_BUILTIN (cv_simd_abs_b_si,            "cv_simd_abs_b",            RISCV_BUILTIN_DIRECT, RISCV_USI_FTYPE_USI, cvsimd),
+RISCV_BUILTIN (cv_simd_dotup_h_si,          "cv_simd_dotup_h",          RISCV_BUILTIN_DIRECT, RISCV_USI_FTYPE_USI_USI, cvsimd),
```
Do you want to insert a comment noting these are the SIMD dot product operations? Makes it easier to reference back to the specification.

jeremybennett replied:
Resolved by newer commits.

jeremybennett commented on gcc/config/riscv/corev.def:
```diff
@@ -20,3 +20,225 @@ RISCV_BUILTIN (cv_mac_macuRN,    "cv_mac_macuRN",     RISCV_BUILTIN_DIRECT, RISC
+RISCV_BUILTIN (cv_simd_abs_b_si,            "cv_simd_abs_b",            RISCV_BUILTIN_DIRECT, RISCV_USI_FTYPE_USI, cvsimd),
+RISCV_BUILTIN (cv_simd_dotup_h_si,          "cv_simd_dotup_h",          RISCV_BUILTIN_DIRECT, RISCV_USI_FTYPE_USI_USI, cvsimd),
+RISCV_BUILTIN (cv_simd_dotup_b_si,          "cv_simd_dotup_b",          RISCV_BUILTIN_DIRECT, RISCV_USI_FTYPE_USI_USI, cvsimd),
+RISCV_BUILTIN (cv_simd_dotup_sc_h_si,       "cv_simd_dotup_sc_h",       RISCV_BUILTIN_DIRECT, RISCV_USI_FTYPE_USI_USI, cvsimd),
```
I think the type should be `RISCV_USI_FTYPE_USI_UHI`

jeremybennett replied:
Resolved by newer commits.

jeremybennett commented on gcc/config/riscv/corev.def:
```diff
@@ -20,3 +20,225 @@ RISCV_BUILTIN (cv_mac_macuRN,    "cv_mac_macuRN",     RISCV_BUILTIN_DIRECT, RISC
+RISCV_BUILTIN (cv_simd_dotup_h_si,          "cv_simd_dotup_h",          RISCV_BUILTIN_DIRECT, RISCV_USI_FTYPE_USI_USI, cvsimd),
+RISCV_BUILTIN (cv_simd_dotup_b_si,          "cv_simd_dotup_b",          RISCV_BUILTIN_DIRECT, RISCV_USI_FTYPE_USI_USI, cvsimd),
+RISCV_BUILTIN (cv_simd_dotup_sc_h_si,       "cv_simd_dotup_sc_h",       RISCV_BUILTIN_DIRECT, RISCV_USI_FTYPE_USI_USI, cvsimd),
+RISCV_BUILTIN (cv_simd_dotup_sc_b_si,       "cv_simd_dotup_sc_b",       RISCV_BUILTIN_DIRECT, RISCV_USI_FTYPE_USI_USI, cvsimd),
```
I think the type should be `RISCV_USI_FTYPE_USI_UQI`

jeremybennett replied:
Resolved by newer commits.

jeremybennett commented on gcc/config/riscv/corev.def:
```diff
@@ -20,3 +20,225 @@ RISCV_BUILTIN (cv_mac_macuRN,    "cv_mac_macuRN",     RISCV_BUILTIN_DIRECT, RISC
+RISCV_BUILTIN (cv_simd_dotup_sci_b_si,      "cv_simd_dotup_sci_b",      RISCV_BUILTIN_DIRECT, RISCV_USI_FTYPE_USI_SI, cvsimd),
+RISCV_BUILTIN (cv_simd_dotusp_h_si,         "cv_simd_dotusp_h",         RISCV_BUILTIN_DIRECT, RISCV_USI_FTYPE_USI_USI, cvsimd),
+RISCV_BUILTIN (cv_simd_dotusp_b_si,         "cv_simd_dotusp_b",         RISCV_BUILTIN_DIRECT, RISCV_USI_FTYPE_USI_USI, cvsimd),
+RISCV_BUILTIN (cv_simd_dotusp_sc_h_si,      "cv_simd_dotusp_sc_h",      RISCV_BUILTIN_DIRECT, RISCV_USI_FTYPE_USI_USI, cvsimd),
```
I think the type should be `RISCV_USI_FTYPE_USI_HI` (the reference is wrong, in this case the last argument is signed).

jeremybennett commented on gcc/config/riscv/corev.def:
```diff
@@ -20,3 +20,225 @@ RISCV_BUILTIN (cv_mac_macuRN,    "cv_mac_macuRN",     RISCV_BUILTIN_DIRECT, RISC
+RISCV_BUILTIN (cv_simd_dotusp_h_si,         "cv_simd_dotusp_h",         RISCV_BUILTIN_DIRECT, RISCV_USI_FTYPE_USI_USI, cvsimd),
+RISCV_BUILTIN (cv_simd_dotusp_b_si,         "cv_simd_dotusp_b",         RISCV_BUILTIN_DIRECT, RISCV_USI_FTYPE_USI_USI, cvsimd),
+RISCV_BUILTIN (cv_simd_dotusp_sc_h_si,      "cv_simd_dotusp_sc_h",      RISCV_BUILTIN_DIRECT, RISCV_USI_FTYPE_USI_USI, cvsimd),
+RISCV_BUILTIN (cv_simd_dotusp_sc_b_si,      "cv_simd_dotusp_sc_b",      RISCV_BUILTIN_DIRECT, RISCV_USI_FTYPE_USI_USI, cvsimd),
```
I think the type should be `RISCV_USI_FTYPE_USI_QI` (the reference is wrong, in this case the last argument is signed).

jeremybennett replied:
Resolved by newer commits.

jeremybennett commented on gcc/config/riscv/corev.def:
```diff
@@ -20,3 +20,225 @@ RISCV_BUILTIN (cv_mac_macuRN,    "cv_mac_macuRN",     RISCV_BUILTIN_DIRECT, RISC
+RISCV_BUILTIN (cv_simd_dotup_sc_h_si,       "cv_simd_dotup_sc_h",       RISCV_BUILTIN_DIRECT, RISCV_USI_FTYPE_USI_USI, cvsimd),
+RISCV_BUILTIN (cv_simd_dotup_sc_b_si,       "cv_simd_dotup_sc_b",       RISCV_BUILTIN_DIRECT, RISCV_USI_FTYPE_USI_USI, cvsimd),
+RISCV_BUILTIN (cv_simd_dotup_sci_h_si,      "cv_simd_dotup_sci_h",      RISCV_BUILTIN_DIRECT, RISCV_USI_FTYPE_USI_SI, cvsimd),
+RISCV_BUILTIN (cv_simd_dotup_sci_b_si,      "cv_simd_dotup_sci_b",      RISCV_BUILTIN_DIRECT, RISCV_USI_FTYPE_USI_SI, cvsimd),
```
As noted earlier the `sci` versions need to go.

jeremybennett replied:
Resolved by newer commits.

jeremybennett commented on gcc/config/riscv/corev.def:
```diff
@@ -20,3 +20,225 @@ RISCV_BUILTIN (cv_mac_macuRN,    "cv_mac_macuRN",     RISCV_BUILTIN_DIRECT, RISC
+RISCV_BUILTIN (cv_simd_dotusp_sci_b_si,     "cv_simd_dotusp_sci_b",     RISCV_BUILTIN_DIRECT, RISCV_USI_FTYPE_USI_SI, cvsimd),
+RISCV_BUILTIN (cv_simd_dotsp_h_si,          "cv_simd_dotsp_h",          RISCV_BUILTIN_DIRECT, RISCV_USI_FTYPE_USI_USI, cvsimd),
+RISCV_BUILTIN (cv_simd_dotsp_b_si,          "cv_simd_dotsp_b",          RISCV_BUILTIN_DIRECT, RISCV_USI_FTYPE_USI_USI, cvsimd),
+RISCV_BUILTIN (cv_simd_dotsp_sc_h_si,       "cv_simd_dotsp_sc_h",       RISCV_BUILTIN_DIRECT, RISCV_USI_FTYPE_USI_USI, cvsimd),
```
I think the type should be `RISCV_USI_FTYPE_USI_HI` (the reference is wrong, in this case the last argument is signed).

jeremybennett replied:
Resolved by newer commits.

jeremybennett commented on gcc/config/riscv/corev.def:
```diff
@@ -20,3 +20,225 @@ RISCV_BUILTIN (cv_mac_macuRN,    "cv_mac_macuRN",     RISCV_BUILTIN_DIRECT, RISC
+RISCV_BUILTIN (cv_simd_dotsp_h_si,          "cv_simd_dotsp_h",          RISCV_BUILTIN_DIRECT, RISCV_USI_FTYPE_USI_USI, cvsimd),
+RISCV_BUILTIN (cv_simd_dotsp_b_si,          "cv_simd_dotsp_b",          RISCV_BUILTIN_DIRECT, RISCV_USI_FTYPE_USI_USI, cvsimd),
+RISCV_BUILTIN (cv_simd_dotsp_sc_h_si,       "cv_simd_dotsp_sc_h",       RISCV_BUILTIN_DIRECT, RISCV_USI_FTYPE_USI_USI, cvsimd),
+RISCV_BUILTIN (cv_simd_dotsp_sc_b_si,       "cv_simd_dotsp_sc_b",       RISCV_BUILTIN_DIRECT, RISCV_USI_FTYPE_USI_USI, cvsimd),
```
I think the type should be `RISCV_USI_FTYPE_USI_QI` (the reference is wrong, in this case the last argument is signed).

jeremybennett replied:
Resolved by newer commits.

jeremybennett commented on gcc/config/riscv/corev.def:
```diff
@@ -20,3 +20,225 @@ RISCV_BUILTIN (cv_mac_macuRN,    "cv_mac_macuRN",     RISCV_BUILTIN_DIRECT, RISC
+RISCV_BUILTIN (cv_simd_dotsp_sc_b_si,       "cv_simd_dotsp_sc_b",       RISCV_BUILTIN_DIRECT, RISCV_USI_FTYPE_USI_USI, cvsimd),
+RISCV_BUILTIN (cv_simd_dotsp_sci_h_si,      "cv_simd_dotsp_sci_h",      RISCV_BUILTIN_DIRECT, RISCV_USI_FTYPE_USI_SI, cvsimd),
+RISCV_BUILTIN (cv_simd_dotsp_sci_b_si,      "cv_simd_dotsp_sci_b",      RISCV_BUILTIN_DIRECT, RISCV_USI_FTYPE_USI_SI, cvsimd),
+RISCV_BUILTIN (cv_simd_sdotup_h_si,         "cv_simd_sdotup_h",         RISCV_BUILTIN_DIRECT, RISCV_USI_FTYPE_USI_USI, cvsimd),
```
Should the type be `RISCV_USI_FTYPE_USI_USI_USI`? Or is this because the final argument is also the destination argument?

jeremybennett replied:
Resolved by newer commits.

jeremybennett commented on gcc/config/riscv/corev.def:
```diff
@@ -20,3 +20,225 @@ RISCV_BUILTIN (cv_mac_macuRN,    "cv_mac_macuRN",     RISCV_BUILTIN_DIRECT, RISC
+RISCV_BUILTIN (cv_simd_dotsp_sci_b_si,      "cv_simd_dotsp_sci_b",      RISCV_BUILTIN_DIRECT, RISCV_USI_FTYPE_USI_SI, cvsimd),
+RISCV_BUILTIN (cv_simd_sdotup_h_si,         "cv_simd_sdotup_h",         RISCV_BUILTIN_DIRECT, RISCV_USI_FTYPE_USI_USI, cvsimd),
+RISCV_BUILTIN (cv_simd_sdotup_b_si,         "cv_simd_sdotup_b",         RISCV_BUILTIN_DIRECT, RISCV_USI_FTYPE_USI_USI, cvsimd),
+RISCV_BUILTIN (cv_simd_sdotup_sc_h_si,      "cv_simd_sdotup_sc_h",      RISCV_BUILTIN_DIRECT, RISCV_USI_FTYPE_USI_USI, cvsimd),
```
I think the type should be `RISCV_USI_FTYPE_USI_UHI_USI`, although the same questions applies as the the half word equivalent.

jeremybennett replied:
Resolved by newer commits.

jeremybennett commented on gcc/config/riscv/corev.def:
```diff
@@ -20,3 +20,225 @@ RISCV_BUILTIN (cv_mac_macuRN,    "cv_mac_macuRN",     RISCV_BUILTIN_DIRECT, RISC
+RISCV_BUILTIN (cv_simd_dotsp_sci_h_si,      "cv_simd_dotsp_sci_h",      RISCV_BUILTIN_DIRECT, RISCV_USI_FTYPE_USI_SI, cvsimd),
+RISCV_BUILTIN (cv_simd_dotsp_sci_b_si,      "cv_simd_dotsp_sci_b",      RISCV_BUILTIN_DIRECT, RISCV_USI_FTYPE_USI_SI, cvsimd),
+RISCV_BUILTIN (cv_simd_sdotup_h_si,         "cv_simd_sdotup_h",         RISCV_BUILTIN_DIRECT, RISCV_USI_FTYPE_USI_USI, cvsimd),
+RISCV_BUILTIN (cv_simd_sdotup_b_si,         "cv_simd_sdotup_b",         RISCV_BUILTIN_DIRECT, RISCV_USI_FTYPE_USI_USI, cvsimd),
```
Againt, should the type be `RISCV_USI_FTYPE_USI_USI_USI`? Or is this because the final argument is also the destination argument?

jeremybennett replied:
Resolved by newer commits.

jeremybennett commented on gcc/config/riscv/corev.def:
```diff
@@ -20,3 +20,225 @@ RISCV_BUILTIN (cv_mac_macuRN,    "cv_mac_macuRN",     RISCV_BUILTIN_DIRECT, RISC
+RISCV_BUILTIN (cv_simd_sdotup_h_si,         "cv_simd_sdotup_h",         RISCV_BUILTIN_DIRECT, RISCV_USI_FTYPE_USI_USI, cvsimd),
+RISCV_BUILTIN (cv_simd_sdotup_b_si,         "cv_simd_sdotup_b",         RISCV_BUILTIN_DIRECT, RISCV_USI_FTYPE_USI_USI, cvsimd),
+RISCV_BUILTIN (cv_simd_sdotup_sc_h_si,      "cv_simd_sdotup_sc_h",      RISCV_BUILTIN_DIRECT, RISCV_USI_FTYPE_USI_USI, cvsimd),
+RISCV_BUILTIN (cv_simd_sdotup_sc_b_si,      "cv_simd_sdotup_sc_b",      RISCV_BUILTIN_DIRECT, RISCV_USI_FTYPE_USI_USI, cvsimd),
```
I think the type should be `RISCV_USI_FTYPE_USI_UQI_USI`, although the same questions apply as the the byte equivalent.

jeremybennett replied:
Resolved by newer commits.

jeremybennett commented on gcc/config/riscv/corev.def:
```diff
@@ -20,3 +20,225 @@ RISCV_BUILTIN (cv_mac_macuRN,    "cv_mac_macuRN",     RISCV_BUILTIN_DIRECT, RISC
+RISCV_BUILTIN (cv_simd_sdotup_b_si,         "cv_simd_sdotup_b",         RISCV_BUILTIN_DIRECT, RISCV_USI_FTYPE_USI_USI, cvsimd),
+RISCV_BUILTIN (cv_simd_sdotup_sc_h_si,      "cv_simd_sdotup_sc_h",      RISCV_BUILTIN_DIRECT, RISCV_USI_FTYPE_USI_USI, cvsimd),
+RISCV_BUILTIN (cv_simd_sdotup_sc_b_si,      "cv_simd_sdotup_sc_b",      RISCV_BUILTIN_DIRECT, RISCV_USI_FTYPE_USI_USI, cvsimd),
+RISCV_BUILTIN (cv_simd_sdotup_sci_h_si,     "cv_simd_sdotup_sci_h",     RISCV_BUILTIN_DIRECT, RISCV_USI_FTYPE_USI_SI, cvsimd),
```
Same comments as for the other dot production intrinsics apply for the rest of this section.

jeremybennett replied:
Resolved by newer commits.

jeremybennett commented on gcc/config/riscv/corev.def:
```diff
@@ -20,3 +20,225 @@ RISCV_BUILTIN (cv_mac_macuRN,    "cv_mac_macuRN",     RISCV_BUILTIN_DIRECT, RISC
+RISCV_BUILTIN (cv_simd_sdotsp_sc_b_si,      "cv_simd_sdotsp_sc_b",      RISCV_BUILTIN_DIRECT, RISCV_USI_FTYPE_USI_USI, cvsimd),
+RISCV_BUILTIN (cv_simd_sdotsp_sci_h_si,     "cv_simd_sdotsp_sci_h",     RISCV_BUILTIN_DIRECT, RISCV_USI_FTYPE_USI_SI, cvsimd),
+RISCV_BUILTIN (cv_simd_sdotsp_sci_b_si,     "cv_simd_sdotsp_sci_b",     RISCV_BUILTIN_DIRECT, RISCV_USI_FTYPE_USI_SI, cvsimd),
+RISCV_BUILTIN (cv_simd_extract_h_si,        "cv_simd_extract_h",        RISCV_BUILTIN_DIRECT, RISCV_USI_FTYPE_USI_QI, cvsimd),
```
A comment before the bitmanip builtins would be useful. Also they should go above the dot products to match the reference documentation.

jeremybennett replied:
Resolved by newer commits.

jeremybennett commented on gcc/config/riscv/corev.def:
```diff
@@ -20,3 +20,225 @@ RISCV_BUILTIN (cv_mac_macuRN,    "cv_mac_macuRN",     RISCV_BUILTIN_DIRECT, RISC
+RISCV_BUILTIN (cv_simd_sdotsp_sc_b_si,      "cv_simd_sdotsp_sc_b",      RISCV_BUILTIN_DIRECT, RISCV_USI_FTYPE_USI_USI, cvsimd),
+RISCV_BUILTIN (cv_simd_sdotsp_sci_h_si,     "cv_simd_sdotsp_sci_h",     RISCV_BUILTIN_DIRECT, RISCV_USI_FTYPE_USI_SI, cvsimd),
+RISCV_BUILTIN (cv_simd_sdotsp_sci_b_si,     "cv_simd_sdotsp_sci_b",     RISCV_BUILTIN_DIRECT, RISCV_USI_FTYPE_USI_SI, cvsimd),
+RISCV_BUILTIN (cv_simd_extract_h_si,        "cv_simd_extract_h",        RISCV_BUILTIN_DIRECT, RISCV_USI_FTYPE_USI_QI, cvsimd),
```
I think in all the following extract intrinsics the type should be `RISCV_USI_FTYPE_USI_UQI`, not `RISCV_USI_FTYPE_USI_QI`

jeremybennett replied:
Resolved by newer commits.

jeremybennett commented on gcc/config/riscv/corev.def:
```diff
@@ -20,3 +20,225 @@ RISCV_BUILTIN (cv_mac_macuRN,    "cv_mac_macuRN",     RISCV_BUILTIN_DIRECT, RISC
+RISCV_BUILTIN (cv_simd_extract_b_si,        "cv_simd_extract_b",        RISCV_BUILTIN_DIRECT, RISCV_USI_FTYPE_USI_QI, cvsimd),
+RISCV_BUILTIN (cv_simd_extractu_h_si,       "cv_simd_extractu_h",       RISCV_BUILTIN_DIRECT, RISCV_USI_FTYPE_USI_QI, cvsimd),
+RISCV_BUILTIN (cv_simd_extractu_b_si,       "cv_simd_extractu_b",       RISCV_BUILTIN_DIRECT, RISCV_USI_FTYPE_USI_QI, cvsimd),
+RISCV_BUILTIN (cv_simd_insert_h_si,         "cv_simd_insert_h",         RISCV_BUILTIN_DIRECT, RISCV_USI_FTYPE_USI_QI, cvsimd),
```
Shouldn't the insert builtins be of type `RISCV_USI_FTYPE_USI_USI_UQI`

jeremybennett replied:
Resolved by newer commits.

jeremybennett commented on gcc/config/riscv/corev.def:
```diff
@@ -20,3 +20,225 @@ RISCV_BUILTIN (cv_mac_macuRN,    "cv_mac_macuRN",     RISCV_BUILTIN_DIRECT, RISC
+RISCV_BUILTIN (cv_simd_insert_b_si,         "cv_simd_insert_b",         RISCV_BUILTIN_DIRECT, RISCV_USI_FTYPE_USI_QI, cvsimd),
+RISCV_BUILTIN (cv_simd_shuffle_h_si,        "cv_simd_shuffle_h",        RISCV_BUILTIN_DIRECT, RISCV_USI_FTYPE_USI_USI, cvsimd),
+RISCV_BUILTIN (cv_simd_shuffle_b_si,        "cv_simd_shuffle_b",        RISCV_BUILTIN_DIRECT, RISCV_USI_FTYPE_USI_USI, cvsimd),
+RISCV_BUILTIN (cv_simd_shuffle_sci_h_si,    "cv_simd_shuffle_sci_h",    RISCV_BUILTIN_DIRECT, RISCV_USI_FTYPE_USI_QI, cvsimd),
```
I think there is a mistake in the documentation, and this should be named `cv_simd_shuffle_sc_h` for consistency.  The same applies for the other `sci` shuffle instructions.  Also I think the type should be `RISCV_USI_FTYPE_USI_UQI`

jeremybennett replied:
Resolved by newer commits.

jeremybennett commented on gcc/config/riscv/corev.def:
```diff
@@ -20,3 +20,225 @@ RISCV_BUILTIN (cv_mac_macuRN,    "cv_mac_macuRN",     RISCV_BUILTIN_DIRECT, RISC
+RISCV_BUILTIN (cv_simd_shuffle_h_si,        "cv_simd_shuffle_h",        RISCV_BUILTIN_DIRECT, RISCV_USI_FTYPE_USI_USI, cvsimd),
+RISCV_BUILTIN (cv_simd_shuffle_b_si,        "cv_simd_shuffle_b",        RISCV_BUILTIN_DIRECT, RISCV_USI_FTYPE_USI_USI, cvsimd),
+RISCV_BUILTIN (cv_simd_shuffle_sci_h_si,    "cv_simd_shuffle_sci_h",    RISCV_BUILTIN_DIRECT, RISCV_USI_FTYPE_USI_QI, cvsimd),
+RISCV_BUILTIN (cv_simd_shuffleI0_sci_b_si,  "cv_simd_shuffleI0_sci_b",  RISCV_BUILTIN_DIRECT, RISCV_USI_FTYPE_USI_QI, cvsimd),
```
There is just one `cv_simd_shuffle_sc_b` function.  The two 2 bits of the constant final argument identify which opcode to generate.  I believe the type should be `RISCV_USI_FTYPE_USI_UQI`

jeremybennett replied:
Resolved by newer commits.

jeremybennett commented on gcc/config/riscv/corev.def:
```diff
@@ -20,3 +20,225 @@ RISCV_BUILTIN (cv_mac_macuRN,    "cv_mac_macuRN",     RISCV_BUILTIN_DIRECT, RISC
+RISCV_BUILTIN (cv_simd_shuffleI1_sci_b_si,  "cv_simd_shuffleI1_sci_b",  RISCV_BUILTIN_DIRECT, RISCV_USI_FTYPE_USI_QI, cvsimd),
+RISCV_BUILTIN (cv_simd_shuffleI2_sci_b_si,  "cv_simd_shuffleI2_sci_b",  RISCV_BUILTIN_DIRECT, RISCV_USI_FTYPE_USI_QI, cvsimd),
+RISCV_BUILTIN (cv_simd_shuffleI3_sci_b_si,  "cv_simd_shuffleI3_sci_b",  RISCV_BUILTIN_DIRECT, RISCV_USI_FTYPE_USI_QI, cvsimd),
+RISCV_BUILTIN (cv_simd_shuffle2_h_si,       "cv_simd_shuffle2_h",       RISCV_BUILTIN_DIRECT, RISCV_USI_FTYPE_USI_USI, cvsimd),
```
Surely the `shuffle2` functions need type `RISCV_USI_FTYPE_USI_USI_USI`?

jeremybennett replied:
Resolved by newer commits.

jeremybennett commented on gcc/config/riscv/corev.def:
```diff
@@ -20,3 +20,225 @@ RISCV_BUILTIN (cv_mac_macuRN,    "cv_mac_macuRN",     RISCV_BUILTIN_DIRECT, RISC
+RISCV_BUILTIN (cv_simd_shuffleI3_sci_b_si,  "cv_simd_shuffleI3_sci_b",  RISCV_BUILTIN_DIRECT, RISCV_USI_FTYPE_USI_QI, cvsimd),
+RISCV_BUILTIN (cv_simd_shuffle2_h_si,       "cv_simd_shuffle2_h",       RISCV_BUILTIN_DIRECT, RISCV_USI_FTYPE_USI_USI, cvsimd),
+RISCV_BUILTIN (cv_simd_shuffle2_b_si,       "cv_simd_shuffle2_b",       RISCV_BUILTIN_DIRECT, RISCV_USI_FTYPE_USI_USI, cvsimd),
+RISCV_BUILTIN (cv_simd_pack_b_si,           "cv_simd_pack_b",           RISCV_BUILTIN_DIRECT, RISCV_USI_FTYPE_USI_USI, cvsimd),
```
There is no `cv_simd_pack_b`. I think this should be `cv_simd_pack`

jeremybennett replied:
Resolved by newer commits.

jeremybennett commented on gcc/config/riscv/corev.def:
```diff
@@ -20,3 +20,225 @@ RISCV_BUILTIN (cv_mac_macuRN,    "cv_mac_macuRN",     RISCV_BUILTIN_DIRECT, RISC
+RISCV_BUILTIN (cv_simd_shuffle2_b_si,       "cv_simd_shuffle2_b",       RISCV_BUILTIN_DIRECT, RISCV_USI_FTYPE_USI_USI, cvsimd),
+RISCV_BUILTIN (cv_simd_pack_b_si,           "cv_simd_pack_b",           RISCV_BUILTIN_DIRECT, RISCV_USI_FTYPE_USI_USI, cvsimd),
+RISCV_BUILTIN (cv_simd_pack_h_si,           "cv_simd_pack_h",           RISCV_BUILTIN_DIRECT, RISCV_USI_FTYPE_USI_USI, cvsimd),
+RISCV_BUILTIN (cv_simd_packhi_b_si,         "cv_simd_packhi_b",         RISCV_BUILTIN_DIRECT, RISCV_USI_FTYPE_USI_USI, cvsimd),
```
I think this and the following instruction are of type `RISCV_USI_FTYPE_USI_USI_USI`

jeremybennett replied:
Resolved by newer commits.

jeremybennett commented on gcc/config/riscv/corev.def:
```diff
@@ -20,3 +20,225 @@ RISCV_BUILTIN (cv_mac_macuRN,    "cv_mac_macuRN",     RISCV_BUILTIN_DIRECT, RISC
+RISCV_BUILTIN (cv_simd_packlo_b_si,         "cv_simd_packlo_b",         RISCV_BUILTIN_DIRECT, RISCV_USI_FTYPE_USI_USI, cvsimd),
+RISCV_BUILTIN (cv_simd_cmpeq_h_si,          "cv_simd_cmpeq_h",          RISCV_BUILTIN_DIRECT, RISCV_USI_FTYPE_USI_USI, cvsimd),
+RISCV_BUILTIN (cv_simd_cmpeq_b_si,          "cv_simd_cmpeq_b",          RISCV_BUILTIN_DIRECT, RISCV_USI_FTYPE_USI_USI, cvsimd),
+RISCV_BUILTIN (cv_simd_cmpeq_sc_h_si,       "cv_simd_cmpeq_sc_h",       RISCV_BUILTIN_DIRECT, RISCV_USI_FTYPE_USI_USI, cvsimd),
```
I think the type should be `RISCV_USI_FTYPE_USI_HI`  (similarly for the other signed half word scalar comparisons)

jeremybennett replied:
Resolved by newer commits.

jeremybennett commented on gcc/config/riscv/corev.def:
```diff
@@ -20,3 +20,225 @@ RISCV_BUILTIN (cv_mac_macuRN,    "cv_mac_macuRN",     RISCV_BUILTIN_DIRECT, RISC
+RISCV_BUILTIN (cv_simd_cmpeq_h_si,          "cv_simd_cmpeq_h",          RISCV_BUILTIN_DIRECT, RISCV_USI_FTYPE_USI_USI, cvsimd),
+RISCV_BUILTIN (cv_simd_cmpeq_b_si,          "cv_simd_cmpeq_b",          RISCV_BUILTIN_DIRECT, RISCV_USI_FTYPE_USI_USI, cvsimd),
+RISCV_BUILTIN (cv_simd_cmpeq_sc_h_si,       "cv_simd_cmpeq_sc_h",       RISCV_BUILTIN_DIRECT, RISCV_USI_FTYPE_USI_USI, cvsimd),
+RISCV_BUILTIN (cv_simd_cmpeq_sc_b_si,       "cv_simd_cmpeq_sc_b",       RISCV_BUILTIN_DIRECT, RISCV_USI_FTYPE_USI_USI, cvsimd),
```
I think the type should be `RISCV_USI_FTYPE_USI_QI`  (similarly for the other signed byte scalar comparisons)

jeremybennett replied:
Resolved by newer commits.

jeremybennett commented on gcc/config/riscv/corev.def:
```diff
@@ -20,3 +20,225 @@ RISCV_BUILTIN (cv_mac_macuRN,    "cv_mac_macuRN",     RISCV_BUILTIN_DIRECT, RISC
+RISCV_BUILTIN (cv_simd_cmple_sci_b_si,      "cv_simd_cmple_sci_b",      RISCV_BUILTIN_DIRECT, RISCV_USI_FTYPE_USI_SI, cvsimd),
+RISCV_BUILTIN (cv_simd_cmpgtu_h_si,         "cv_simd_cmpgtu_h",         RISCV_BUILTIN_DIRECT, RISCV_USI_FTYPE_USI_USI, cvsimd),
+RISCV_BUILTIN (cv_simd_cmpgtu_b_si,         "cv_simd_cmpgtu_b",         RISCV_BUILTIN_DIRECT, RISCV_USI_FTYPE_USI_USI, cvsimd),
+RISCV_BUILTIN (cv_simd_cmpgtu_sc_h_si,      "cv_simd_cmpgtu_sc_h",      RISCV_BUILTIN_DIRECT, RISCV_USI_FTYPE_USI_USI, cvsimd),
```
I think the type should be `RISCV_USI_FTYPE_USI_UHI`  (similarly for the other unsigned half word scalar comparisons)

jeremybennett replied:
Resolved by newer commits.

jeremybennett commented on gcc/config/riscv/corev.def:
```diff
@@ -20,3 +20,225 @@ RISCV_BUILTIN (cv_mac_macuRN,    "cv_mac_macuRN",     RISCV_BUILTIN_DIRECT, RISC
+RISCV_BUILTIN (cv_simd_cmpgtu_h_si,         "cv_simd_cmpgtu_h",         RISCV_BUILTIN_DIRECT, RISCV_USI_FTYPE_USI_USI, cvsimd),
+RISCV_BUILTIN (cv_simd_cmpgtu_b_si,         "cv_simd_cmpgtu_b",         RISCV_BUILTIN_DIRECT, RISCV_USI_FTYPE_USI_USI, cvsimd),
+RISCV_BUILTIN (cv_simd_cmpgtu_sc_h_si,      "cv_simd_cmpgtu_sc_h",      RISCV_BUILTIN_DIRECT, RISCV_USI_FTYPE_USI_USI, cvsimd),
+RISCV_BUILTIN (cv_simd_cmpgtu_sc_b_si,      "cv_simd_cmpgtu_sc_b",      RISCV_BUILTIN_DIRECT, RISCV_USI_FTYPE_USI_USI, cvsimd),
```
I think the type should be `RISCV_USI_FTYPE_USI_UQI`  (similarly for the other unsigned byte scalar comparisons)

jeremybennett replied:
Resolved by newer commits.

jeremybennett commented on gcc/config/riscv/corev.def:
```diff
@@ -20,3 +20,225 @@ RISCV_BUILTIN (cv_mac_macuRN,    "cv_mac_macuRN",     RISCV_BUILTIN_DIRECT, RISC
+RISCV_BUILTIN (cv_simd_cmpleu_sc_b_si,      "cv_simd_cmpleu_sc_b",      RISCV_BUILTIN_DIRECT, RISCV_USI_FTYPE_USI_USI, cvsimd),
+RISCV_BUILTIN (cv_simd_cmpleu_sci_h_si,     "cv_simd_cmpleu_sci_h",     RISCV_BUILTIN_DIRECT, RISCV_USI_FTYPE_USI_SI, cvsimd),
+RISCV_BUILTIN (cv_simd_cmpleu_sci_b_si,     "cv_simd_cmpleu_sci_b",     RISCV_BUILTIN_DIRECT, RISCV_USI_FTYPE_USI_SI, cvsimd),
+RISCV_BUILTIN (cv_simd_cplxmul_r_si,        "cv_simd_cplxmul_r",        RISCV_BUILTIN_DIRECT, RISCV_USI_FTYPE_USI_USI, cvsimd),
```
Type should be `RISCV_USI_FTYPE_USI_USI_USI_UQI`, with the final argument used to control which/wether the `.div` variants of the instruction are generated.

jeremybennett replied:
Resolved by newer commits.

jeremybennett commented on gcc/config/riscv/corev.def:
```diff
@@ -20,3 +20,225 @@ RISCV_BUILTIN (cv_mac_macuRN,    "cv_mac_macuRN",     RISCV_BUILTIN_DIRECT, RISC
+RISCV_BUILTIN (cv_simd_cmpleu_sci_h_si,     "cv_simd_cmpleu_sci_h",     RISCV_BUILTIN_DIRECT, RISCV_USI_FTYPE_USI_SI, cvsimd),
+RISCV_BUILTIN (cv_simd_cmpleu_sci_b_si,     "cv_simd_cmpleu_sci_b",     RISCV_BUILTIN_DIRECT, RISCV_USI_FTYPE_USI_SI, cvsimd),
+RISCV_BUILTIN (cv_simd_cplxmul_r_si,        "cv_simd_cplxmul_r",        RISCV_BUILTIN_DIRECT, RISCV_USI_FTYPE_USI_USI, cvsimd),
+RISCV_BUILTIN (cv_simd_cplxmul_i_si,        "cv_simd_cplxmul_i",        RISCV_BUILTIN_DIRECT, RISCV_USI_FTYPE_USI_USI, cvsimd),
```
Type should be `RISCV_USI_FTYPE_USI_USI_USI_UQI`, with the final argument used to control which/wether the `.div` variants of the instruction are generated.

jeremybennett replied:
Resolved by newer commits.

jeremybennett commented on gcc/config/riscv/corev.def:
```diff
@@ -20,3 +20,225 @@ RISCV_BUILTIN (cv_mac_macuRN,    "cv_mac_macuRN",     RISCV_BUILTIN_DIRECT, RISC
+RISCV_BUILTIN (cv_simd_cmpleu_sci_b_si,     "cv_simd_cmpleu_sci_b",     RISCV_BUILTIN_DIRECT, RISCV_USI_FTYPE_USI_SI, cvsimd),
+RISCV_BUILTIN (cv_simd_cplxmul_r_si,        "cv_simd_cplxmul_r",        RISCV_BUILTIN_DIRECT, RISCV_USI_FTYPE_USI_USI, cvsimd),
+RISCV_BUILTIN (cv_simd_cplxmul_i_si,        "cv_simd_cplxmul_i",        RISCV_BUILTIN_DIRECT, RISCV_USI_FTYPE_USI_USI, cvsimd),
+RISCV_BUILTIN (cv_simd_cplxmul_r_div2_si,   "cv_simd_cplxmul_r_div2",   RISCV_BUILTIN_DIRECT, RISCV_USI_FTYPE_USI_USI, cvsimd),
```
There are no `simd_cplxmul_?_div?` builtins. The `.div?` versions of instrructions are generate as appropriate by the `simd_cplxmul_?` builtin.

jeremybennett replied:
Resolved by newer commits.

jeremybennett commented on gcc/config/riscv/corev.def:
```diff
@@ -20,3 +20,225 @@ RISCV_BUILTIN (cv_mac_macuRN,    "cv_mac_macuRN",     RISCV_BUILTIN_DIRECT, RISC
+RISCV_BUILTIN (cv_simd_subrotmj_div2_si,    "cv_simd_subrotmj_div2",    RISCV_BUILTIN_DIRECT, RISCV_USI_FTYPE_USI_USI, cvsimd),
+RISCV_BUILTIN (cv_simd_subrotmj_div4_si,    "cv_simd_subrotmj_div4",    RISCV_BUILTIN_DIRECT, RISCV_USI_FTYPE_USI_USI, cvsimd),
+RISCV_BUILTIN (cv_simd_subrotmj_div8_si,    "cv_simd_subrotmj_div8",    RISCV_BUILTIN_DIRECT, RISCV_USI_FTYPE_USI_USI, cvsimd),
+RISCV_BUILTIN (cv_simd_add_div2_si,         "cv_simd_add_div2",         RISCV_BUILTIN_DIRECT, RISCV_USI_FTYPE_USI_USI, cvsimd),
```
Home of the `cv_simd_add/sub_div` builtins are needed.  The relevant instructions are generated as needed by the `cv_simd_add_h` and `cv_simd_sub_h` builtins

jeremybennett replied:
Resolved by newer commits.

jeremybennett commented on gcc/config/riscv/corev.def:
```diff
@@ -20,3 +20,225 @@ RISCV_BUILTIN (cv_mac_macuRN,    "cv_mac_macuRN",     RISCV_BUILTIN_DIRECT, RISC
+RISCV_BUILTIN (cv_simd_cplxmul_r_div8_si,   "cv_simd_cplxmul_r_div8",   RISCV_BUILTIN_DIRECT, RISCV_USI_FTYPE_USI_USI, cvsimd),
+RISCV_BUILTIN (cv_simd_cplxmul_i_div8_si,   "cv_simd_cplxmul_i_div8",   RISCV_BUILTIN_DIRECT, RISCV_USI_FTYPE_USI_USI, cvsimd),
+RISCV_BUILTIN (cv_simd_cplxconj_si,         "cv_simd_cplxconj",         RISCV_BUILTIN_DIRECT, RISCV_USI_FTYPE_USI, cvsimd),
+RISCV_BUILTIN (cv_simd_subrotmj_si,         "cv_simd_subrotmj",         RISCV_BUILTIN_DIRECT, RISCV_USI_FTYPE_USI_USI, cvsimd),
```
This should be of type `RISCV_USI_FTYPE_USI_USI_UQI`, with the final constant argument used to determine whether to generate one of the `_div` variants.

jeremybennett replied:
Resolved by newer commits.

jeremybennett commented on gcc/config/riscv/corev.def:
```diff
@@ -20,3 +20,225 @@ RISCV_BUILTIN (cv_mac_macuRN,    "cv_mac_macuRN",     RISCV_BUILTIN_DIRECT, RISC
+RISCV_BUILTIN (cv_simd_cplxmul_i_div8_si,   "cv_simd_cplxmul_i_div8",   RISCV_BUILTIN_DIRECT, RISCV_USI_FTYPE_USI_USI, cvsimd),
+RISCV_BUILTIN (cv_simd_cplxconj_si,         "cv_simd_cplxconj",         RISCV_BUILTIN_DIRECT, RISCV_USI_FTYPE_USI, cvsimd),
+RISCV_BUILTIN (cv_simd_subrotmj_si,         "cv_simd_subrotmj",         RISCV_BUILTIN_DIRECT, RISCV_USI_FTYPE_USI_USI, cvsimd),
+RISCV_BUILTIN (cv_simd_subrotmj_div2_si,    "cv_simd_subrotmj_div2",    RISCV_BUILTIN_DIRECT, RISCV_USI_FTYPE_USI_USI, cvsimd),
```
There are no `cv_simd_subrotmj_div?` builtins.

jeremybennett replied:
Resolved by newer commits.

jeremybennett commented on gcc/config/riscv/corev.def:
```diff
@@ -20,3 +20,225 @@ RISCV_BUILTIN (cv_mac_macuRN,    "cv_mac_macuRN",     RISCV_BUILTIN_DIRECT, RISC
+// xcoresimd
+RISCV_BUILTIN (cv_simd_add_h_si,            "cv_simd_add_h",            RISCV_BUILTIN_DIRECT, RISCV_USI_FTYPE_USI_USI, cvsimd),
+RISCV_BUILTIN (cv_simd_add_b_si,            "cv_simd_add_b",            RISCV_BUILTIN_DIRECT, RISCV_USI_FTYPE_USI_USI, cvsimd),
+RISCV_BUILTIN (cv_simd_add_sc_h_si,         "cv_simd_add_sc_h",         RISCV_BUILTIN_DIRECT, RISCV_USI_FTYPE_USI_USI, cvsimd),
```
I think the type shoudl be `RISCV_USI_FTYPE_USI_HI`

jeremybennett replied:
Resolved by newer commits.

jeremybennett commented on gcc/config/riscv/corev.def:
```diff
@@ -20,3 +20,225 @@ RISCV_BUILTIN (cv_mac_macuRN,    "cv_mac_macuRN",     RISCV_BUILTIN_DIRECT, RISC
+RISCV_BUILTIN (cv_simd_add_h_si,            "cv_simd_add_h",            RISCV_BUILTIN_DIRECT, RISCV_USI_FTYPE_USI_USI, cvsimd),
+RISCV_BUILTIN (cv_simd_add_b_si,            "cv_simd_add_b",            RISCV_BUILTIN_DIRECT, RISCV_USI_FTYPE_USI_USI, cvsimd),
+RISCV_BUILTIN (cv_simd_add_sc_h_si,         "cv_simd_add_sc_h",         RISCV_BUILTIN_DIRECT, RISCV_USI_FTYPE_USI_USI, cvsimd),
+RISCV_BUILTIN (cv_simd_add_sc_b_si,         "cv_simd_add_sc_b",         RISCV_BUILTIN_DIRECT, RISCV_USI_FTYPE_USI_USI, cvsimd),
```
I think the type shoudl be `RISCV_USI_FTYPE_USI_QI`

jeremybennett replied:
Resolved by newer commits.

jeremybennett reviewed (changes requested):
This is an interim review, to capture various issues, which need addressing.

NandniJamnadas committed:
CV32E40Pv2 220 CORE-V SIMD Builtins Version 1

    * gcc/config/riscv/constraints.md: Created
        * 6-bit signed immediate constraint "CV6"
        * 6-bit unsigned immediate constraint "CS6"
        * 1-bit shifting immediate constraint:
          * constant integer value "0" constraint "CF0"
          * constant integer value "1" constraint "CF1"
          * constant integer value "2" constraint "CF2"
          * constant integer value "3" constraint "CF3"

    * gcc/config/riscv/corev.def: Defined 142 SIMD builtins
      for 220 SIMD instructions with RISCV_BUILTIN.

    * gcc/config/riscv/corev.md:
      * Machine description pattern for all 142 SIMD builtins with
        UNSPEC version.
      * Included `define_c_enum "unspec"` for CORE-V specific
        unspec macros.

    * gcc/config/riscv/predicates.md: Created
        * 6-bit signed immediate predicate for constraint "CV6"
        * 6-bit unsigned immediate predicate for constraint "CS6"
        * 4-bit unsigned immediate predicate for constraint
          "CF0,CF1,CF2,CF3"

    * gcc/config/riscv/riscv-builtins.cc: Created SIMD Avaliability
      Predicate, RISCV_FTYPE_NAME2, RISCV_FTYPE_ATYPES2. Added
      corresponding argument types for builtins.

    * gcc/config/riscv/riscv-ftypes.def: Added corresponding
      DEF_RISCV_FTYPE return type and argument types for builtins.

NandniJamnadas committed:
CV32E40Pv2 220 CORE-V SIMD Builtins: Compilation Tests

	158 Compilation Tests and 1 MARCH test: cv-xcorevsimd-march-compile-1.c

	* gcc/testsuite/gcc.target/riscv/cv-simd-abs-b-compile-1.c: Created.
	* gcc/testsuite/gcc.target/riscv/cv-simd-abs-h-compile-1.c: Likewise.
	* gcc/testsuite/gcc.target/riscv/cv-simd-add-b-compile-1.c: Likewise.
	* gcc/testsuite/gcc.target/riscv/cv-simd-add-div2-compile-1.c: Likewise.
	* gcc/testsuite/gcc.target/riscv/cv-simd-add-div4-compile-1.c: Likewise.
	* gcc/testsuite/gcc.target/riscv/cv-simd-add-div8-compile-1.c: Likewise.
	* gcc/testsuite/gcc.target/riscv/cv-simd-add-h-compile-1.c: Likewise.
	* gcc/testsuite/gcc.target/riscv/cv-simd-add-sc-b-compile-1.c: Likewise.
	* gcc/testsuite/gcc.target/riscv/cv-simd-add-sc-h-compile-1.c: Likewise.
	* gcc/testsuite/gcc.target/riscv/cv-simd-and-b-compile-1.c: Likewise.
	* gcc/testsuite/gcc.target/riscv/cv-simd-and-h-compile-1.c: Likewise.
	* gcc/testsuite/gcc.target/riscv/cv-simd-and-sc-b-compile-1.c: Likewise.
	* gcc/testsuite/gcc.target/riscv/cv-simd-and-sc-h-compile-1.c: Likewise.
	* gcc/testsuite/gcc.target/riscv/cv-simd-avg-b-compile-1.c: Likewise.
	* gcc/testsuite/gcc.target/riscv/cv-simd-avg-h-compile-1.c: Likewise.
	* gcc/testsuite/gcc.target/riscv/cv-simd-avg-sc-b-compile-1.c: Likewise.
	* gcc/testsuite/gcc.target/riscv/cv-simd-avg-sc-h-compile-1.c: Likewise.
	* gcc/testsuite/gcc.target/riscv/cv-simd-avgu-b-compile-1.c: Likewise.
	* gcc/testsuite/gcc.target/riscv/cv-simd-avgu-h-compile-1.c: Likewise.
	* gcc/testsuite/gcc.target/riscv/cv-simd-avgu-sc-b-compile-1.c: Likewise.
	* gcc/testsuite/gcc.target/riscv/cv-simd-avgu-sc-h-compile-1.c: Likewise.
	* gcc/testsuite/gcc.target/riscv/cv-simd-cmpeq-b-compile-1.c: Likewise.
	* gcc/testsuite/gcc.target/riscv/cv-simd-cmpeq-h-compile-1.c: Likewise.
	* gcc/testsuite/gcc.target/riscv/cv-simd-cmpeq-sc-b-compile-1.c: Likewise.
	* gcc/testsuite/gcc.target/riscv/cv-simd-cmpeq-sc-h-compile-1.c: Likewise.
	* gcc/testsuite/gcc.target/riscv/cv-simd-cmpge-b-compile-1.c: Likewise.
	* gcc/testsuite/gcc.target/riscv/cv-simd-cmpge-h-compile-1.c: Likewise.
	* gcc/testsuite/gcc.target/riscv/cv-simd-cmpge-sc-b-compile-1.c: Likewise.
	* gcc/testsuite/gcc.target/riscv/cv-simd-cmpge-sc-h-compile-1.c: Likewise.
	* gcc/testsuite/gcc.target/riscv/cv-simd-cmpgeu-b-compile-1.c: Likewise.
	* gcc/testsuite/gcc.target/riscv/cv-simd-cmpgeu-h-compile-1.c: Likewise.
	* gcc/testsuite/gcc.target/riscv/cv-simd-cmpgeu-sc-b-compile-1.c: Likewise.
	* gcc/testsuite/gcc.target/riscv/cv-simd-cmpgeu-sc-h-compile-1.c: Likewise.
	* gcc/testsuite/gcc.target/riscv/cv-simd-cmpgt-b-compile-1.c: Likewise.
	* gcc/testsuite/gcc.target/riscv/cv-simd-cmpgt-h-compile-1.c: Likewise.
	* gcc/testsuite/gcc.target/riscv/cv-simd-cmpgt-sc-b-compile-1.c: Likewise.
	* gcc/testsuite/gcc.target/riscv/cv-simd-cmpgt-sc-h-compile-1.c: Likewise.
	* gcc/testsuite/gcc.target/riscv/cv-simd-cmpgtu-b-compile-1.c: Likewise.
	* gcc/testsuite/gcc.target/riscv/cv-simd-cmpgtu-h-compile-1.c: Likewise.
	* gcc/testsuite/gcc.target/riscv/cv-simd-cmpgtu-sc-b-compile-1.c: Likewise.
	* gcc/testsuite/gcc.target/riscv/cv-simd-cmpgtu-sc-h-compile-1.c: Likewise.
	* gcc/testsuite/gcc.target/riscv/cv-simd-cmple-b-compile-1.c: Likewise.
	* gcc/testsuite/gcc.target/riscv/cv-simd-cmple-h-compile-1.c: Likewise.
	* gcc/testsuite/gcc.target/riscv/cv-simd-cmple-sc-b-compile-1.c: Likewise.
	* gcc/testsuite/gcc.target/riscv/cv-simd-cmple-sc-h-compile-1.c: Likewise.
	* gcc/testsuite/gcc.target/riscv/cv-simd-cmpleu-b-compile-1.c: Likewise.
	* gcc/testsuite/gcc.target/riscv/cv-simd-cmpleu-h-compile-1.c: Likewise.
	* gcc/testsuite/gcc.target/riscv/cv-simd-cmpleu-sc-b-compile-1.c: Likewise.
	* gcc/testsuite/gcc.target/riscv/cv-simd-cmpleu-sc-h-compile-1.c: Likewise.
	* gcc/testsuite/gcc.target/riscv/cv-simd-cmplt-b-compile-1.c: Likewise.
	* gcc/testsuite/gcc.target/riscv/cv-simd-cmplt-h-compile-1.c: Likewise.
	* gcc/testsuite/gcc.target/riscv/cv-simd-cmplt-sc-b-compile-1.c: Likewise.
	* gcc/testsuite/gcc.target/riscv/cv-simd-cmplt-sc-h-compile-1.c: Likewise.
	* gcc/testsuite/gcc.target/riscv/cv-simd-cmpltu-b-compile-1.c: Likewise.
	* gcc/testsuite/gcc.target/riscv/cv-simd-cmpltu-h-compile-1.c: Likewise.
	* gcc/testsuite/gcc.target/riscv/cv-simd-cmpltu-sc-b-compile-1.c: Likewise.
	* gcc/testsuite/gcc.target/riscv/cv-simd-cmpltu-sc-h-compile-1.c: Likewise.
	* gcc/testsuite/gcc.target/riscv/cv-simd-cmpne-b-compile-1.c: Likewise.
	* gcc/testsuite/gcc.target/riscv/cv-simd-cmpne-h-compile-1.c: Likewise.
	* gcc/testsuite/gcc.target/riscv/cv-simd-cmpne-sc-b-compile-1.c: Likewise.
	* gcc/testsuite/gcc.target/riscv/cv-simd-cmpne-sc-h-compile-1.c: Likewise.
	* gcc/testsuite/gcc.target/riscv/cv-simd-cplxconj-compile-1.c: Likewise.
	* gcc/testsuite/gcc.target/riscv/cv-simd-cplxmul-i-compile-1.c: Likewise.
	* gcc/testsuite/gcc.target/riscv/cv-simd-cplxmul-i-div2-compile-1.c: Likewise.
	* gcc/testsuite/gcc.target/riscv/cv-simd-cplxmul-i-div4-compile-1.c: Likewise.
	* gcc/testsuite/gcc.target/riscv/cv-simd-cplxmul-i-div8-compile-1.c: Likewise.
	* gcc/testsuite/gcc.target/riscv/cv-simd-cplxmul-r-compile-1.c: Likewise.
	* gcc/testsuite/gcc.target/riscv/cv-simd-cplxmul-r-div2-compile-1.c: Likewise.
	* gcc/testsuite/gcc.target/riscv/cv-simd-cplxmul-r-div4-compile-1.c: Likewise.
	* gcc/testsuite/gcc.target/riscv/cv-simd-cplxmul-r-div8-compile-1.c: Likewise.
	* gcc/testsuite/gcc.target/riscv/cv-simd-dotsp-b-compile-1.c: Likewise.
	* gcc/testsuite/gcc.target/riscv/cv-simd-dotsp-h-compile-1.c: Likewise.
	* gcc/testsuite/gcc.target/riscv/cv-simd-dotsp-sc-b-compile-1.c: Likewise.
	* gcc/testsuite/gcc.target/riscv/cv-simd-dotsp-sc-h-compile-1.c: Likewise.
	* gcc/testsuite/gcc.target/riscv/cv-simd-dotup-b-compile-1.c: Likewise.
	* gcc/testsuite/gcc.target/riscv/cv-simd-dotup-h-compile-1.c: Likewise.
	* gcc/testsuite/gcc.target/riscv/cv-simd-dotup-sc-b-compile-1.c: Likewise.
	* gcc/testsuite/gcc.target/riscv/cv-simd-dotup-sc-h-compile-1.c: Likewise.
	* gcc/testsuite/gcc.target/riscv/cv-simd-dotusp-b-compile-1.c: Likewise.
	* gcc/testsuite/gcc.target/riscv/cv-simd-dotusp-h-compile-1.c: Likewise.
	* gcc/testsuite/gcc.target/riscv/cv-simd-dotusp-sc-b-compile-1.c: Likewise.
	* gcc/testsuite/gcc.target/riscv/cv-simd-dotusp-sc-h-compile-1.c: Likewise.
	* gcc/testsuite/gcc.target/riscv/cv-simd-extract-b-compile-1.c: Likewise.
	* gcc/testsuite/gcc.target/riscv/cv-simd-extract-h-compile-1.c: Likewise.
	* gcc/testsuite/gcc.target/riscv/cv-simd-extractu-b-compile-1.c: Likewise.
	* gcc/testsuite/gcc.target/riscv/cv-simd-extractu-h-compile-1.c: Likewise.
	* gcc/testsuite/gcc.target/riscv/cv-simd-insert-b-compile-1.c: Likewise.
	* gcc/testsuite/gcc.target/riscv/cv-simd-insert-h-compile-1.c: Likewise.
	* gcc/testsuite/gcc.target/riscv/cv-simd-max-b-compile-1.c: Likewise.
	* gcc/testsuite/gcc.target/riscv/cv-simd-max-h-compile-1.c: Likewise.
	* gcc/testsuite/gcc.target/riscv/cv-simd-max-sc-b-compile-1.c: Likewise.
	* gcc/testsuite/gcc.target/riscv/cv-simd-max-sc-h-compile-1.c: Likewise.
	* gcc/testsuite/gcc.target/riscv/cv-simd-maxu-b-compile-1.c: Likewise.
	* gcc/testsuite/gcc.target/riscv/cv-simd-maxu-h-compile-1.c: Likewise.
	* gcc/testsuite/gcc.target/riscv/cv-simd-maxu-sc-b-compile-1.c: Likewise.
	* gcc/testsuite/gcc.target/riscv/cv-simd-maxu-sc-h-compile-1.c: Likewise.
	* gcc/testsuite/gcc.target/riscv/cv-simd-min-b-compile-1.c: Likewise.
	* gcc/testsuite/gcc.target/riscv/cv-simd-min-h-compile-1.c: Likewise.
	* gcc/testsuite/gcc.target/riscv/cv-simd-min-sc-b-compile-1.c: Likewise.
	* gcc/testsuite/gcc.target/riscv/cv-simd-min-sc-h-compile-1.c: Likewise.
	* gcc/testsuite/gcc.target/riscv/cv-simd-minu-b-compile-1.c: Likewise.
	* gcc/testsuite/gcc.target/riscv/cv-simd-minu-h-compile-1.c: Likewise.
	* gcc/testsuite/gcc.target/riscv/cv-simd-minu-sc-b-compile-1.c: Likewise.
	* gcc/testsuite/gcc.target/riscv/cv-simd-minu-sc-h-compile-1.c: Likewise.
	* gcc/testsuite/gcc.target/riscv/cv-simd-or-b-compile-1.c: Likewise.
	* gcc/testsuite/gcc.target/riscv/cv-simd-or-h-compile-1.c: Likewise.
	* gcc/testsuite/gcc.target/riscv/cv-simd-or-sc-b-compile-1.c: Likewise.
	* gcc/testsuite/gcc.target/riscv/cv-simd-or-sc-h-compile-1.c: Likewise.
	* gcc/testsuite/gcc.target/riscv/cv-simd-pack-compile-1.c: Likewise.
	* gcc/testsuite/gcc.target/riscv/cv-simd-pack-h-compile-1.c: Likewise.
	* gcc/testsuite/gcc.target/riscv/cv-simd-packhi-b-compile-1.c: Likewise.
	* gcc/testsuite/gcc.target/riscv/cv-simd-packlo-b-compile-1.c: Likewise.
	* gcc/testsuite/gcc.target/riscv/cv-simd-sdotsp-b-compile-1.c: Likewise.
	* gcc/testsuite/gcc.target/riscv/cv-simd-sdotsp-h-compile-1.c: Likewise.
	* gcc/testsuite/gcc.target/riscv/cv-simd-sdotsp-sc-b-compile-1.c: Likewise.
	* gcc/testsuite/gcc.target/riscv/cv-simd-sdotsp-sc-h-compile-1.c: Likewise.
	* gcc/testsuite/gcc.target/riscv/cv-simd-sdotup-b-compile-1.c: Likewise.
	* gcc/testsuite/gcc.target/riscv/cv-simd-sdotup-h-compile-1.c: Likewise.
	* gcc/testsuite/gcc.target/riscv/cv-simd-sdotup-sc-b-compile-1.c: Likewise.
	* gcc/testsuite/gcc.target/riscv/cv-simd-sdotup-sc-h-compile-1.c: Likewise.
	* gcc/testsuite/gcc.target/riscv/cv-simd-sdotusp-b-compile-1.c: Likewise.
	* gcc/testsuite/gcc.target/riscv/cv-simd-sdotusp-h-compile-1.c: Likewise.
	* gcc/testsuite/gcc.target/riscv/cv-simd-sdotusp-sc-b-compile-1.c: Likewise.
	* gcc/testsuite/gcc.target/riscv/cv-simd-sdotusp-sc-h-compile-1.c: Likewise.
	* gcc/testsuite/gcc.target/riscv/cv-simd-shuffle-sci-h-compile-1.c: Likewise.
	* gcc/testsuite/gcc.target/riscv/cv-simd-shuffle2-b-compile-1.c: Likewise.
	* gcc/testsuite/gcc.target/riscv/cv-simd-shuffle2-h-compile-1.c: Likewise.
	* gcc/testsuite/gcc.target/riscv/cv-simd-shufflei0-sci-b-compile-1.c: Likewise.
	* gcc/testsuite/gcc.target/riscv/cv-simd-shufflei1-sci-b-compile-1.c: Likewise.
	* gcc/testsuite/gcc.target/riscv/cv-simd-shufflei2-sci-b-compile-1.c: Likewise.
	* gcc/testsuite/gcc.target/riscv/cv-simd-shufflei3-sci-b-compile-1.c: Likewise.
	* gcc/testsuite/gcc.target/riscv/cv-simd-sll-b-compile-1.c: Likewise.
	* gcc/testsuite/gcc.target/riscv/cv-simd-sll-h-compile-1.c: Likewise.
	* gcc/testsuite/gcc.target/riscv/cv-simd-sll-sc-b-compile-1.c: Likewise.
	* gcc/testsuite/gcc.target/riscv/cv-simd-sll-sc-h-compile-1.c: Likewise.
	* gcc/testsuite/gcc.target/riscv/cv-simd-sra-b-compile-1.c: Likewise.
	* gcc/testsuite/gcc.target/riscv/cv-simd-sra-h-compile-1.c: Likewise.
	* gcc/testsuite/gcc.target/riscv/cv-simd-sra-sc-b-compile-1.c: Likewise.
	* gcc/testsuite/gcc.target/riscv/cv-simd-sra-sc-h-compile-1.c: Likewise.
	* gcc/testsuite/gcc.target/riscv/cv-simd-srl-b-compile-1.c: Likewise.
	* gcc/testsuite/gcc.target/riscv/cv-simd-srl-h-compile-1.c: Likewise.
	* gcc/testsuite/gcc.target/riscv/cv-simd-srl-sc-b-compile-1.c: Likewise.
	* gcc/testsuite/gcc.target/riscv/cv-simd-srl-sc-h-compile-1.c: Likewise.
	* gcc/testsuite/gcc.target/riscv/cv-simd-sub-b-compile-1.c: Likewise.
	* gcc/testsuite/gcc.target/riscv/cv-simd-sub-div2-compile-1.c: Likewise.
	* gcc/testsuite/gcc.target/riscv/cv-simd-sub-div4-compile-1.c: Likewise.
	* gcc/testsuite/gcc.target/riscv/cv-simd-sub-div8-compile-1.c: Likewise.
	* gcc/testsuite/gcc.target/riscv/cv-simd-sub-h-compile-1.c: Likewise.
	* gcc/testsuite/gcc.target/riscv/cv-simd-sub-sc-b-compile-1.c: Likewise.
	* gcc/testsuite/gcc.target/riscv/cv-simd-sub-sc-h-compile-1.c: Likewise.
	* gcc/testsuite/gcc.target/riscv/cv-simd-subrotmj-compile-1.c: Likewise.
	* gcc/testsuite/gcc.target/riscv/cv-simd-subrotmj-div2-compile-1.c: Likewise.
	* gcc/testsuite/gcc.target/riscv/cv-simd-subrotmj-div4-compile-1.c: Likewise.
	* gcc/testsuite/gcc.target/riscv/cv-simd-subrotmj-div8-compile-1.c: Likewise.
	* gcc/testsuite/gcc.target/riscv/cv-simd-xor-b-compile-1.c: Likewise.
	* gcc/testsuite/gcc.target/riscv/cv-simd-xor-h-compile-1.c: Likewise.
	* gcc/testsuite/gcc.target/riscv/cv-simd-xor-sc-b-compile-1.c: Likewise.
	* gcc/testsuite/gcc.target/riscv/cv-simd-xor-sc-h-compile-1.c: Likewise.
	* gcc/testsuite/gcc.target/riscv/cv-xcorevsimd-march-compile-1.c: Likewise.

jeremybennett commented on gcc/config/riscv/riscv-ftypes.def:
```diff
@@ -28,9 +28,21 @@ along with GCC; see the file COPYING3.  If not see
+DEF_RISCV_FTYPE (3, (USI, USI, USI, UQI))       //RISCV_USI_FTYPE_USI_USI_UQI
+DEF_RISCV_FTYPE (3, (SI, SI, SI, UQI))          //RISCV_SI_FTYPE_SI_SI_UQI
+DEF_RISCV_FTYPE (4, (USI, USI, USI, USI, UQI))  //RISCV_USI_FTYPE_USI_USI_USI_UQI
+DEF_RISCV_FTYPE (4, (SI, SI, SI, SI, UQI))      //RISCV_SI_FTYPE_SI_SI_SI_UQI
```
Note warning about missing newline at end of file! Check other files as well (looks to be all tests!).

jeremybennett replied:
All fixed.

jeremybennett commented on gcc/testsuite/gcc.target/riscv/cv-simd-subrotmj-div2-compile-1.c:
```diff
@@ -0,0 +1,9 @@
+/* { dg-do compile } */
```
I would be minded to merge all the subrotmj tests into one and have multiple scan-assembler lines at the end.  No need to change now, but would be a good cleanup later.

jeremybennett reviewed (changes requested):
Great piece of work.  There is a piece of technical cleanup needed to add in all the missing newlines at end of file. The other comment is a suggestion for the future.

With the missing newlines added, this will be good to merge.

jeremybennett commented on gcc/config/riscv/corev.def:
```diff
@@ -20,3 +20,153 @@ RISCV_BUILTIN (cv_mac_macuRN,    "cv_mac_macuRN",     RISCV_BUILTIN_DIRECT, RISC
+
+// XCOREVSIMD
+//ALU Operations
+RISCV_BUILTIN (cv_simd_add_h_si,			"cv_simd_add_h",     RISCV_BUILTIN_DIRECT, RISCV_USI_FTYPE_USI_USI_UQI, cvsimd),
```
Need to fix layout - possibly a tabbing issue?

jeremybennett replied:
Better, but still looks a bit odd in GitHub.  I suggest we leave this for a future cleanup.

jeremybennett commented on gcc/config/riscv/corev.def:
```diff
@@ -20,3 +20,153 @@ RISCV_BUILTIN (cv_mac_macuRN,    "cv_mac_macuRN",     RISCV_BUILTIN_DIRECT, RISC
+RISCV_BUILTIN (cv_simd_abs_h_si,			"cv_simd_abs_h",     RISCV_BUILTIN_DIRECT, RISCV_USI_FTYPE_USI, cvsimd),
+RISCV_BUILTIN (cv_simd_abs_b_si,			"cv_simd_abs_b",     RISCV_BUILTIN_DIRECT, RISCV_USI_FTYPE_USI, cvsimd),
+//BIT MANIPULATION
+RISCV_BUILTIN (cv_simd_extract_h_si,		"cv_simd_extract_h",     RISCV_BUILTIN_DIRECT, RISCV_USI_FTYPE_USI_QI, cvsimd),
```
This should be `RISCV_SI_FTYPE_USI_QI` as per the documentation

jeremybennett replied:
All fixed.

jeremybennett commented on gcc/config/riscv/corev.def:
```diff
@@ -20,3 +20,153 @@ RISCV_BUILTIN (cv_mac_macuRN,    "cv_mac_macuRN",     RISCV_BUILTIN_DIRECT, RISC
+RISCV_BUILTIN (cv_simd_abs_b_si,			"cv_simd_abs_b",     RISCV_BUILTIN_DIRECT, RISCV_USI_FTYPE_USI, cvsimd),
+//BIT MANIPULATION
+RISCV_BUILTIN (cv_simd_extract_h_si,		"cv_simd_extract_h",     RISCV_BUILTIN_DIRECT, RISCV_USI_FTYPE_USI_QI, cvsimd),
+RISCV_BUILTIN (cv_simd_extract_b_si,		"cv_simd_extract_b",     RISCV_BUILTIN_DIRECT, RISCV_USI_FTYPE_USI_QI, cvsimd),
```
This should be `RISCV_SI_FTYPE_USI_QI` as per the documentation

jeremybennett replied:
All fixed

jeremybennett reviewed (changes requested):
Spotted a couple of type errors.

NandniJamnadas committed:
CV32E40Pv2 220 CORE-V SIMD Builtins: Documentation

	* gcc/doc/extend.texi: Documentation on 142 SIMD Built-ins
	  for 220 SIMD instructions.

jeremybennett reviewed (approved):
All issues now addressed.  Good to merge.
